src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -6271,229 +6271,205 @@ private TypeSyntax ParsePointerTypeMods(TypeSyntax type)
+                    case SyntaxKind.ForKeyword:
+                        return this.ParseForOrForEachStatement();
+                    case SyntaxKind.ForEachKeyword:
+                        return this.ParseForEachStatement(awaitTokenOpt: default);
+                    case SyntaxKind.GotoKeyword:
+                        return this.ParseGotoStatement();
+                    case SyntaxKind.IfKeyword:
+                        return this.ParseIfStatement();
+                    case SyntaxKind.ElseKeyword:
+                        // Including 'else' keyword to handle 'else without if' error cases 
+                        return this.ParseMisplacedElse();
+                    case SyntaxKind.LockKeyword:
+                        return this.ParseLockStatement();
+                    case SyntaxKind.ReturnKeyword:
+                        return this.ParseReturnStatement();
+                    case SyntaxKind.SwitchKeyword:
+                        return this.ParseSwitchStatement();
+                    case SyntaxKind.ThrowKeyword:
+                        return this.ParseThrowStatement();
+                    case SyntaxKind.UnsafeKeyword:
+                        result = TryParseStatementStartingWithUnsafe();
+                        if (result != null)
+                            return result;
+                        break;
+                    case SyntaxKind.UsingKeyword:
+                        return ParseStatementStartingWithUsing();
+                    case SyntaxKind.WhileKeyword:
+                        return this.ParseWhileStatement();
+                    case SyntaxKind.OpenBraceToken:
+                        return this.ParseBlock();
+                    case SyntaxKind.SemicolonToken:
+                        return _syntaxFactory.EmptyStatement(this.EatToken());
+                    case SyntaxKind.IdentifierToken:
+                        result = TryParseStatementStartingWithIdentifier(isGlobalScriptLevel);
+                        if (result != null)
+                            return result;
+                        break;
                 }
 
-                // We could not successfully parse the statement as a non-declaration. Try to parse
-                // it as either a declaration or as an "await X();" statement that is in a non-async
-                // method. 
-
-                return TryParsePossibleDeclarationOrBadAwaitStatement();
+                return ParseStatementCoreRest(isGlobalScriptLevel, ref resetPointBeforeStatement);
             }
             finally
             {
                 _recursionDepth--;
+                this.Release(ref resetPointBeforeStatement);
+            }
+
+            bool canReuseStatement(bool isGlobalScriptLevel)
+            {
+                // Can only reuse a global-script or regular statement if we're in the same
+                // global-script context we were originally in.
+                return this.IsIncrementalAndFactoryContextMatches &&
```

Comment:
I feel the code is clearer with this check done inline instead of with the local function.